base/base/getMemoryAmount.cpp

```diff
@@ -16,26 +16,30 @@
   */
 uint64_t getMemoryAmountOrZero()
 {
+    int64_t num_pages = sysconf(_SC_PHYS_PAGES);
+    if (num_pages <= 0)
+        return 0;
+
+    int64_t page_size = getPageSize();
+    if (page_size <= 0)
+        return 0;
+
+    uint64_t memory_amount = num_pages * page_size;
+
 #if defined(OS_LINUX)
     // Try to lookup at the Cgroup limit
     std::ifstream cgroup_limit("/sys/fs/cgroup/memory/memory.limit_in_bytes");
     if (cgroup_limit.is_open())
     {
-        uint64_t amount = 0; // in case of read error
-        cgroup_limit >> amount;
-        return amount;
+        uint64_t memory_limit = 0; // in case of read error
+        cgroup_limit >> memory_limit;
+        if (memory_limit > 0 && memory_limit < memory_amount)
```

Comment:
If you are using std streams, consider adding `cgroup_limit.good()` check.
Or even better, consider converting `std::ifstream` to `DB::ReadBufferFromFile`